app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/EnqueuedView.kt

```diff
@@ -1,77 +1,219 @@
 package me.hufman.androidautoidrive.carapp.music.views
 
 import de.bmw.idrive.BMWRemoting
+import me.hufman.androidautoidrive.GraphicsHelpers
 import me.hufman.androidautoidrive.UnicodeCleaner
 import me.hufman.androidautoidrive.carapp.RHMIListAdapter
 import me.hufman.androidautoidrive.carapp.music.MusicImageIDs
 import me.hufman.androidautoidrive.music.MusicController
 import me.hufman.androidautoidrive.music.MusicMetadata
+import me.hufman.androidautoidrive.music.QueueMetadata
 import me.hufman.idriveconnectionkit.rhmi.*
+import kotlin.math.max
 
-class EnqueuedView(val state: RHMIState, val musicController: MusicController, val musicImageIDs: MusicImageIDs) {
+class EnqueuedView(val state: RHMIState, val musicController: MusicController, val graphicsHelpers: GraphicsHelpers, val musicImageIDs: MusicImageIDs) {
 	companion object {
+		//current default row width only supports 22 chars before rolling over
+		private const val ROW_LINE_MAX_LENGTH = 22
+
 		fun fits(state: RHMIState): Boolean {
 			return state is RHMIState.PlainState &&
 					state.componentsList.filterIsInstance<RHMIComponent.List>().isNotEmpty() &&
-					state.componentsList.filterIsInstance<RHMIComponent.Image>().isEmpty() &&
+					state.componentsList.filterIsInstance<RHMIComponent.Image>().isNotEmpty() &&
 					state.componentsList.filterIsInstance<RHMIComponent.Separator>().isEmpty()
 		}
 	}
 
 	val listComponent: RHMIComponent.List
+	val queueImageComponent: RHMIComponent.Image
+	val titleLabelComponent: RHMIComponent.Label
+	val subtitleLabelComponent: RHMIComponent.Label
 	var currentSong: MusicMetadata? = null
 	val songsList = ArrayList<MusicMetadata>()
 	val songsEmptyList = RHMIModel.RaListModel.RHMIListConcrete(3)
-	val songsListAdapter = object: RHMIListAdapter<MusicMetadata>(3, songsList) {
+	var queueMetadata: QueueMetadata? = null
+	var visibleRows: List<MusicMetadata> = emptyList()
+	var visibleRowsOriginalMusicMetadata: ArrayList<MusicMetadata> = ArrayList()
+	var selectedIndex: Int = 0
+
+	var songsListAdapter = object: RHMIListAdapter<MusicMetadata>(4, songsList) {
 		override fun convertRow(index: Int, item: MusicMetadata): Array<Any> {
 			val checkmark = if (item.queueId == currentSong?.queueId) BMWRemoting.RHMIResourceIdentifier(BMWRemoting.RHMIResourceType.IMAGEID, musicImageIDs.CHECKMARK) else ""
-			return arrayOf(checkmark,
-						"",
-						UnicodeCleaner.clean(item.title ?: ""))
+
+			val coverArtImage = if (item.coverArt != null) graphicsHelpers.compress(item.coverArt!!, 90, 90, quality = 30) else ""
```

Comment:
There is a miniscule chance that the `item`, perhaps a SpotifyMusicMetadata, might have its coverart disappear after the `if !null` check, because of the underlying LRUCache. To be absolutely safe, save the `item.coverArt` to a local variable and check/compress that.

Reply:
That's true, I didn't consider that. Updated to store the cover art in a local var